nncf/quantization/algorithms/hyperparameter_tuner/algorithm.py

```diff
@@ -0,0 +1,365 @@
+# Key of combination without the trailing zeros
+TrimedCombinationKey = Tuple[int, ...]
+Combination = Dict[str, Any]
+
+
+def create_combinations(param_settings: Dict[str, List[Any]]) -> Dict[CombinationKey, Combination]:
+    """
+    Creates combination of parameters.
+
+    To create key of combination (CombinationKey) used the following approach
+
+    -- All keys in `param_settings` are numbered used zero-based indices i.e.
+    0 - for first key, 1 - for second and so on.
+    -- If `combination_key[i]` is `0` it means that value of parameter i was not changed.
+    In other words, `0` means that initial value (from `init_params`) of parameter is used.
+    -- Let index i corresponds `param_name` key from `param_settings`. If `combination_key[i]`
+    is not `0`, then `combination_key[i]` is a position of value in `param_settings[param_name]`
+    that was included to combination.
+
+    :param param_settings: Dictionary with parameters names as keys and list of
+        parameter settings to try as values.
+    :return: Created combinations.
+    """
+    simple_changes = [[{}, *[{param_name: v} for v in values]] for param_name, values in param_settings.items()]
+
+    combinations = {}
+    for num_params in range(1, len(simple_changes) + 1):
+        for group in itertools.product(*map(enumerate, simple_changes[:num_params])):
+            combination_key, members = zip(*group)
+
+            combination = {}
+            for m in members:
+                combination.update(m)
+            combinations[combination_key] = combination
+
+    return combinations
+
+
+def is_dataclass_instance(obj: Any) -> bool:
+    """
+    Returns `True` if object is a dataclass instance, `False` otherwise.
+
+    :param obj: Object to check.
+    :return: `True` if object is a dataclass instance, `False` otherwise.
+    """
+    return dataclasses.is_dataclass(obj) and not isinstance(obj, type)
+
+
+def apply_combination(init_params: Dict[str, Any], combination: Combination) -> Dict[str, Any]:
+    """
+    Applies combination of parameters to initial parameters.
+
+    :param init_params: Initial set of parameters.
+    :param combination: Combination of parameters.
+    :return: Returns `init_params` where some values of parameters were changed according to
+        provided combination.
+    """
+    DELIMITER = ":"
+    params = copy.deepcopy(init_params)
+    for param_key, param_value in combination.items():
+        if DELIMITER in param_key:
+            main_key, *path_to_attr, attr_name = param_key.split(DELIMITER)
+            obj = params[main_key]
+            assert is_dataclass_instance(obj)
+            for name in path_to_attr:
+                obj = getattr(obj, name)
+                assert is_dataclass_instance(obj)
+            setattr(obj, attr_name, param_value)
+        else:
+            params[param_key] = param_value
+
+    return params
+
+
+def trim_zeros(t: Tuple[int, ...]) -> Tuple[int, ...]:
```

Comment:
This method uses everywhere. Can it be explained with examples, what precisely does it do?

Reply:
The `trim_zeros()` was removed.